File: frontend/src/liveSeries.ts
import type { ChannelParams, Measurement } from './api'

export type SeriesPoint = { t: number; v: number | null; i: number | null; c: number | null }

export type ServerChannelSeries = {
  t0: number
  points: SeriesPoint[]
}

type ChannelSeries = {
  t0: number
  points: SeriesPoint[]
}

const seriesByChannel = new Map<number, ChannelSeries>()
const listeners = new Set<() => void>()

let channelsSnap: ChannelParams[] = []
let measurementsSnap: Measurement[] = []
let intervalId: number | null = null
/** Empty live payloads (disconnect race) before accepting an empty snap. */
let emptySnapStreak = 0
const EMPTY_SNAP_STREAK = 3

type PendingSample = { v: number | null; i: number | null; c: number | null }
type MetricKey = 'v' | 'i' | 'c'
type MetricPending = { value: number | null; confirms: number }
type PendingState = Partial<Record<MetricKey, MetricPending>>
const pendingByChannel = new Map<number, PendingState>()
/** After Stop, do not start a new series until the channel is seen idle. */
const holdOffChannels = new Set<number>()
const readoutHold = new Map<number, PendingSample>()
const readoutPending = new Map<number, PendingState>()

const GLITCH_V_ABS = 0.15 // V
const GLITCH_V_REL = 0.04
const GLITCH_V_FLOOR = 0.05
const GLITCH_I_ABS = 60 // mA
const GLITCH_I_REL = 0.15
const GLITCH_I_FLOOR = 10
const GLITCH_C_ABS = 20 // mAh
const GLITCH_C_REL = 0.03
const GLITCH_C_FLOOR = 5
/** Confirms before accepting a non-collapse abrupt step (ramps, stage changes). */
const CONFIRM_DEFAULT = 3
/** Absolute capacity above this is treated as wire garbage. */
const CAPACITY_ABSURD = 15000
/** Scrub collapse/spike runs up to this many seconds between two stable neighbors. */
const SCRUB_RUN_MAX = 20
/** Trickle/top-off is often 50–100 mA; 120 mA hid those plateaus from collapse-hold. */
const STABLE_I_MA = 20

function isIdle(c: ChannelParams | undefined): boolean {
  if (!c) return true
  return Boolean(c.idle) || c.stage_name === 'Leerlauf'
}

function notify(): void {
  for (const listener of listeners) {
    try {
      listener()
    } catch {
      /* one bad subscriber must not stop other charts */
    }
  }
}

function dropLegacySessionKeys(): void {
  if (typeof sessionStorage === 'undefined') return
  try {
    for (const key of [
      'alc-live-series-v1',
      'alc-live-series-v2',
      'alc-live-series-v3',
      'alc-live-series-v4',
      'alc-live-series-v5',
      'alc-live-series-v6',
    ]) {
      sessionStorage.removeItem(key)
    }
  } catch {
    /* private mode */
  }
}

dropLegacySessionKeys()

function clearChannel(channel: number, silent = false): void {
  pendingByChannel.delete(channel)
  if (!seriesByChannel.has(channel)) return
  seriesByChannel.delete(channel)
  if (!silent) notify()
}

export function clearSeries(channel: number, holdOff = false): void {
  if (holdOff) holdOffChannels.add(channel)
  else holdOffChannels.delete(channel)
  readoutHold.delete(channel)
  readoutPending.delete(channel)
  clearChannel(channel)
}

export function clearAllSeries(): void {
  pendingByChannel.clear()
  holdOffChannels.clear()
  readoutHold.clear()
  readoutPending.clear()
  if (seriesByChannel.size === 0) return
  seriesByChannel.clear()
  notify()
}

/** Replace local charts from dashboard-process RAM (host/origin independent). */
export function replaceFromServer(channels: Record<string, ServerChannelSeries> | null | undefined): void {
  pendingByChannel.clear()
  seriesByChannel.clear()
  if (channels && typeof channels === 'object') {
    for (const [key, value] of Object.entries(channels)) {
      const ch = Number(key)
      if (!Number.isFinite(ch) || !value?.points || !Array.isArray(value.points)) continue
      if (holdOffChannels.has(ch)) continue
      const points = holdCollapses(scrubPoints(value.points.filter((p) => p && typeof p.t === 'number')))
      if (points.length === 0) continue
      seriesByChannel.set(ch, {
        t0: typeof value.t0 === 'number' ? value.t0 : Date.now(),
        points,
      })
    }
  }
  notify()
}

export function getSeries(channel: number): SeriesPoint[] {
  const pts = seriesByChannel.get(channel)?.points
  return pts ? holdCollapses(scrubPoints(pts)) : []
}

export function subscribe(listener: () => void): () => void {
  listeners.add(listener)
  return () => {
    listeners.delete(listener)
  }
}

function anyRunning(): boolean {
  return channelsSnap.some((c) => !isIdle(c))
}

function metricAbrupt(
  prev: number | null | undefined,
  next: number | null | undefined,
  key: MetricKey,
): boolean {
  if (prev == null || next == null) {
    return (prev == null) !== (next == null)
  }
  return Math.abs(prev - next) >= threshFor(key, prev)
}

function threshFor(key: MetricKey, ref?: number | null): number {
  const abs = key === 'v' ? GLITCH_V_ABS : key === 'i' ? GLITCH_I_ABS : GLITCH_C_ABS
  const rel = key === 'v' ? GLITCH_V_REL : key === 'i' ? GLITCH_I_REL : GLITCH_C_REL
  const floor = key === 'v' ? GLITCH_V_FLOOR : key === 'i' ? GLITCH_I_FLOOR : GLITCH_C_FLOOR
  if (ref == null || !Number.isFinite(ref)) return abs
  return Math.max(floor, Math.min(abs, rel * Math.abs(ref)))
}

function sanitizeRaw(raw: PendingSample): PendingSample {
  let { v, i, c } = raw
  if (c != null && (c < 0 || c > CAPACITY_ABSURD)) c = null
  if (i != null && Math.abs(i) > 20000) i = null
  if (v != null && (v < -1 || v > 80)) v = null
  return { v, i, c }
}

/** True if U, I, or C jumped harder than a plausible 1 Hz process step. */
export function isAbruptSample(
  prev: SeriesPoint,
  v: number | null,
  i: number | null,
  c: number | null,
): boolean {
  return (
    metricAbrupt(prev.v, v, 'v') ||
    metricAbrupt(prev.i, i, 'i') ||
    metricAbrupt(prev.c, c, 'c')
  )
}

function nearMetric(a: number | null | undefined, b: number | null | undefined, key: MetricKey): boolean {
  return !metricAbrupt(a, b, key)
}

/** Near-zero / missing reading for a metric (the needle tip). */
export function isCollapsedValue(value: number | null | undefined, key: MetricKey): boolean {
  if (value == null) return true
  if (key === 'i') return Math.abs(value) < 40
  if (key === 'c') return value < 5
  return value < 0.25
}

/** Previously healthy level that should not snap to the axis. */
export function isStableValue(value: number | null | undefined, key: MetricKey): boolean {
  if (value == null) return false
  if (key === 'i') return Math.abs(value) >= STABLE_I_MA
  if (key === 'c') return value >= 30
  return value >= 1.2
}

/**
 * Stable → ~0/null collapse (U/I/C needles). Never accept into the series while running.
 */
export function isMetricCollapse(
  key: MetricKey,
  prev: number | null | undefined,
  next: number | null | undefined,
): boolean {
  return isStableValue(prev, key) && isCollapsedValue(next, key)
}

export function isCurrentCollapse(prevI: number | null | undefined, nextI: number | null | undefined): boolean {
  return isMetricCollapse('i', prevI, nextI)
}

export function isCapacityCollapse(prevC: number | null | undefined, nextC: number | null | undefined): boolean {
  return isMetricCollapse('c', prevC, nextC)
}

/**
 * Replace collapse/spike runs where both neighbors sit on the same plateau.
 * Never spreads collapsed zeros into good data (left must not be collapsed).
 */
export function scrubPoints(points: SeriesPoint[]): SeriesPoint[] {
  if (points.length < 3) return points
  const out = points.map((p) => ({ ...p }))
  const keys: MetricKey[] = ['v', 'i', 'c']

  for (const key of keys) {
    for (let j = 1; j < out.length - 1; j++) {
      if (nearMetric(out[j - 1][key], out[j + 1][key], key) && metricAbrupt(out[j - 1][key], out[j][key], key)) {
        out[j] = { ...out[j], [key]: out[j - 1][key] }
      }
    }

    let i = 1
    while (i < out.length - 1) {
      if (isCollapsedValue(out[i - 1][key], key)) {
        i += 1
        continue
      }
      if (!metricAbrupt(out[i - 1][key], out[i][key], key)) {
        i += 1
        continue
      }

      const runStart = i
      while (i < out.length && metricAbrupt(out[runStart - 1][key], out[i][key], key)) {
        i += 1
      }
      const runLen = i - runStart
      if (runLen < 1 || runLen > SCRUB_RUN_MAX || i >= out.length) continue

      const left = out[runStart - 1][key]
      const right = out[i][key]
      if (isCollapsedValue(left, key) || !nearMetric(left, right, key)) continue

      // Prefer filling collapses; also patch brief non-collapse spikes between plateaus.
      let fill = true
      const collapseRun = Array.from({ length: runLen }, (_, j) =>
        isCollapsedValue(out[runStart + j][key], key),
      ).every(Boolean)
      if (!collapseRun) {
        // Non-collapse spike: every point must be abrupt vs both neighbors' plateau.
        for (let j = 0; j < runLen; j++) {
          if (!metricAbrupt(left, out[runStart + j][key], key)) {
            fill = false
            break
          }
        }
      }
      if (!fill) continue

      for (let j = 0; j < runLen; j++) {
        const t = (j + 1) / (runLen + 1)
        const interpolated =
          left != null && right != null ? left + (right - left) * t : left
        out[runStart + j] = { ...out[runStart + j], [key]: interpolated }
      }
    }
  }
  return out
}

/** Stable → ~0/null keeps the last healthy level (including trailing needles). */
export function holdCollapses(points: SeriesPoint[]): SeriesPoint[] {
  if (points.length < 2) return points
  const out = points.map((p) => ({ ...p }))
  const keys: MetricKey[] = ['v', 'i', 'c']
  for (let i = 1; i < out.length; i++) {
    for (const key of keys) {
      if (isMetricCollapse(key, out[i - 1][key], out[i][key])) {
        out[i] = { ...out[i], [key]: out[i - 1][key] }
      }
    }
  }
  return out
}

function resolveMetric(
  pending: PendingState,
  key: MetricKey,
  prev: number | null,
  raw: number | null,
): number | null {
  // Collapses (stable → ~0/null) are never written while the channel is running.
  if (isMetricCollapse(key, prev, raw)) {
    pending[key] = { value: raw, confirms: (pending[key]?.confirms ?? 0) + 1 }
    return prev
  }

  const slot = pending[key]
  if (slot) {
    if (isMetricCollapse(key, prev, slot.value)) {
      // Pending collapse: resume only when raw is back near the stable prev.
      if (nearMetric(prev, raw, key)) {
        delete pending[key]
        return raw
      }
      pending[key] = { value: raw, confirms: slot.confirms + 1 }
      return prev
    }

    if (nearMetric(slot.value, raw, key)) {
      const confirms = slot.confirms + 1
      if (confirms >= CONFIRM_DEFAULT) {
        delete pending[key]
        return raw
      }
      pending[key] = { value: raw, confirms }
      return prev
    }
    if (nearMetric(prev, raw, key)) {
      delete pending[key]
      return raw
    }
    pending[key] = { value: raw, confirms: 1 }
    return prev
  }

  if (metricAbrupt(prev, raw, key)) {
    pending[key] = { value: raw, confirms: 1 }
    return prev
  }

  return raw
}

function resolveSampleWith(
  pendingMap: Map<number, PendingState>,
  channel: number,
  prev: PendingSample | SeriesPoint | null,
  incoming: PendingSample,
): PendingSample {
  const raw = sanitizeRaw(incoming)
  if (!prev) {
    pendingMap.delete(channel)
    return raw
  }

  let pending = pendingMap.get(channel)
  if (!pending) {
    pending = {}
    pendingMap.set(channel, pending)
  }

  const sample: PendingSample = {
    v: resolveMetric(pending, 'v', prev.v, raw.v),
    i: resolveMetric(pending, 'i', prev.i, raw.i),
    c: resolveMetric(pending, 'c', prev.c, raw.c),
  }

  if (Object.keys(pending).length === 0) {
    pendingMap.delete(channel)
  }

  return sample
}

/**
 * Hold abrupt metrics until enough matching ticks (filters multi-second needles).
 * Metrics are independent so a current glitch does not freeze voltage/capacity.
 */
function resolveSample(
  channel: number,
  prev: SeriesPoint | null,
  incoming: PendingSample,
): PendingSample {
  return resolveSampleWith(pendingByChannel, channel, prev, incoming)
}

/** Smooth live U/I/C readouts with the same hold/scrub rules as the chart. */
export function smoothLiveMeasurements(
  channels: ChannelParams[],
  measurements: Measurement[],
): Measurement[] {
  const chList = channels.map((c, idx) => ({ ...c, channel: idx }))
  const measList =
    measurements.length === chList.length
      ? measurements.map((m, idx) => ({ ...m, channel: idx }))
      : measurements

  return measList.map((m) => {
    const ch = m.channel
    const c = chList.find((x) => x.channel === ch)
    if (isIdle(c)) {
      readoutHold.delete(ch)
      readoutPending.delete(ch)
      return m
    }
    const incoming = { v: m.voltage_V ?? null, i: m.current_mA ?? null, c: m.capacity_mAh ?? null }
    const prev = readoutHold.get(ch) ?? null
    const sample = resolveSampleWith(readoutPending, ch, prev, incoming)
    readoutHold.set(ch, sample)
    return {
      ...m,
      voltage_V: sample.v,
      current_mA: sample.i,
      capacity_mAh: sample.c,
    }
  })
}

function appendPoint(series: ChannelSeries, now: number, sample: PendingSample): void {
  series.points.push({
    t: (now - series.t0) / 1000,
    v: sample.v,
    i: sample.i,
    c: sample.c,
  })
  // Keep stored history free of sandwich spikes (wider window than before)
  if (series.points.length >= 3) {
    const n = series.points.length
    const window = Math.min(n, SCRUB_RUN_MAX * 2 + 3)
    const scrubbed = holdCollapses(scrubPoints(series.points.slice(-window)))
    series.points = [...series.points.slice(0, Math.max(0, n - window)), ...scrubbed]
  }
}

function syncTimer(): void {
  if (typeof window === 'undefined') return
  const running = anyRunning()
  if (running && intervalId == null) {
    sampleTick()
    intervalId = window.setInterval(sampleTick, 1000)
  } else if (!running && intervalId != null) {
    window.clearInterval(intervalId)
    intervalId = null
  }
}

function sampleTick(): void {
  let changed = false
  const now = Date.now()

  if (channelsSnap.length === 0) {
    return
  }

  for (const chParams of channelsSnap) {
    const ch = chParams.channel
    if (isIdle(chParams)) {
      holdOffChannels.delete(ch)
      continue
    }
    if (holdOffChannels.has(ch)) continue

    const m = measurementsSnap.find((x) => x.channel === ch)
    let series = seriesByChannel.get(ch)
    if (!series) {
      series = { t0: now, points: [] }
      seriesByChannel.set(ch, series)
    }

    const raw = {
      v: m?.voltage_V ?? null,
      i: m?.current_mA ?? null,
      c: m?.capacity_mAh ?? null,
    }
    const prev = series.points.length > 0 ? series.points[series.points.length - 1] : null
    const sample = resolveSample(ch, prev, raw)
    appendPoint(series, now, sample)
    changed = true
  }

  if (changed) notify()

  if (!anyRunning() && intervalId != null) {
    window.clearInterval(intervalId)
    intervalId = null
  }
}

/** Feed latest live channels/measurements; samples once per second while any channel is running. */
export function updateLiveSnapshot(channels: ChannelParams[], measurements: Measurement[]): void {
  if (channels.length === 0) {
    emptySnapStreak += 1
    if (emptySnapStreak < EMPTY_SNAP_STREAK) {
      syncTimer()
      return
    }
  } else {
    emptySnapStreak = 0
  }

  channelsSnap = channels.map((c, idx) => ({ ...c, channel: idx }))
  measurementsSnap =
    measurements.length === channelsSnap.length
      ? measurements.map((m, idx) => ({ ...m, channel: idx }))
      : measurements

  syncTimer()
}

File: backend/app/services/live_series.py
"""In-process live U/I/C series — last 24 hours per channel, RAM only."""

from __future__ import annotations

import math
import threading
import time
from collections import deque
from typing import Any, Literal

MAX_AGE_S = 24 * 3600
MAX_POINTS = 24 * 3600  # 1 Hz cap
_MAX_AGE_MS = MAX_AGE_S * 1000

MetricKey = Literal["v", "i", "c"]
Point = tuple[int, float, float | None, float | None, float | None]
PendingSlot = dict[str, Any]
PendingState = dict[str, PendingSlot]

# Match frontend/src/liveSeries.ts — hold serial glitches instead of drawing axis needles.
# Absolute caps apply at high levels; relative band catches trickle/top-off needles
# (e.g. 93 mA → 45 mA) that stay well below the 60 mA absolute jump.
GLITCH_V_ABS = 0.15  # V
GLITCH_V_REL = 0.04
GLITCH_V_FLOOR = 0.05
GLITCH_I_ABS = 60.0  # mA
GLITCH_I_REL = 0.15
GLITCH_I_FLOOR = 10.0
GLITCH_C_ABS = 20.0  # mAh
GLITCH_C_REL = 0.03
GLITCH_C_FLOOR = 5.0
CONFIRM_DEFAULT = 3
CAPACITY_ABSURD = 200000.0  # 200 Ah — wire garbage, not a 1 s process step
# Capacity barely moves at 1 Hz (~mAh/s). Jumps beyond this are serial spikes, never accept.
CAPACITY_JUMP_ABS = 80.0
CAPACITY_JUMP_REL = 0.2
SCRUB_RUN_MAX = 20
STABLE_I_MA = 20.0  # trickle/top-off is often 50–100 mA; 120 hid those plateaus
# After Stop/Start, ignore leftover high C until a fresh low baseline or timeout.
CAPACITY_FRESH_TIMEOUT_MS = 15_000


def _now_ms(now_ms: int | None) -> int:
    if now_ms is not None:
        return int(now_ms)
    return int(time.time() * 1000)


def is_idle_channel(ch: dict[str, Any] | None) -> bool:
    if not ch:
        return True
    return bool(ch.get("idle")) or ch.get("stage_name") == "Leerlauf"


def _finite(value: Any) -> float | None:
    if value is None:
        return None
    try:
        x = float(value)
    except (TypeError, ValueError):
        return None
    if not math.isfinite(x):
        return None
    return x


def thresh_for(key: MetricKey, ref: float | None = None) -> float:
    if key == "v":
        abs_t, rel, floor = GLITCH_V_ABS, GLITCH_V_REL, GLITCH_V_FLOOR
    elif key == "i":
        abs_t, rel, floor = GLITCH_I_ABS, GLITCH_I_REL, GLITCH_I_FLOOR
    else:
        abs_t, rel, floor = GLITCH_C_ABS, GLITCH_C_REL, GLITCH_C_FLOOR
    if ref is None or not math.isfinite(ref):
        return abs_t
    return max(floor, min(abs_t, rel * abs(ref)))


def metric_abrupt(prev: float | None, nxt: float | None, key: MetricKey) -> bool:
    if prev is None or nxt is None:
        return (prev is None) != (nxt is None)
    return abs(prev - nxt) >= thresh_for(key, prev)


def is_collapsed_value(value: float | None, key: MetricKey) -> bool:
    if value is None:
        return True
    if key == "i":
        return abs(value) < 40.0
    if key == "c":
        return value < 5.0
    return value < 0.25


def is_stable_value(value: float | None, key: MetricKey) -> bool:
    if value is None:
        return False
    if key == "i":
        return abs(value) >= STABLE_I_MA
    if key == "c":
        return value >= 30.0
    return value >= 1.2


def is_implausible_capacity_jump(prev: float | None, nxt: float | None) -> bool:
    """True if C changed faster than any real 1 Hz process (e.g. 5000 → 1_000_000)."""
    if prev is None or not math.isfinite(prev):
        return False
    if nxt is None or not math.isfinite(nxt):
        return is_stable_value(prev, "c")
    return abs(nxt - prev) >= max(CAPACITY_JUMP_ABS, CAPACITY_JUMP_REL * abs(prev))


def is_capacity_down_reset(prev: float | None, nxt: float | None) -> bool:
    """New process / counter reset: capacity dropped hard (not an upward serial spike)."""
    if not is_implausible_capacity_jump(prev, nxt):
        return False
    if prev is None or not math.isfinite(prev):
        return False
    if nxt is None or not math.isfinite(nxt):
        return is_stable_value(prev, "c")
    return nxt < prev


def is_metric_collapse(key: MetricKey, prev: float | None, nxt: float | None) -> bool:
    if is_stable_value(prev, key) and is_collapsed_value(nxt, key):
        return True
    if key == "c":
        return is_implausible_capacity_jump(prev, nxt)
    return False


def near_metric(a: float | None, b: float | None, key: MetricKey) -> bool:
    return not metric_abrupt(a, b, key)


def sanitize_raw(
    voltage: Any, current: Any, capacity: Any
) -> tuple[float | None, float | None, float | None]:
    v = _finite(voltage)
    i = _finite(current)
    c = _finite(capacity)
    if c is not None and (c < 0 or c > CAPACITY_ABSURD):
        c = None
    if i is not None and abs(i) > 20000:
        i = None
    if v is not None and (v < -1 or v > 80):
        v = None
    return v, i, c


def resolve_metric(
    pending: PendingState,
    key: MetricKey,
    prev: float | None,
    raw: float | None,
) -> float | None:
    # V/I collapses stay held. Capacity down-resets unlock after confirmed ticks;
    # upward capacity spikes never unlock.
    if is_metric_collapse(key, prev, raw):
        slot = pending.get(key)
        confirms = int(slot["confirms"]) + 1 if slot else 1
        if key == "c" and is_capacity_down_reset(prev, raw) and confirms >= CONFIRM_DEFAULT:
            pending.pop(key, None)
            return raw
        pending[key] = {"value": raw, "confirms": confirms}
        return prev

    slot = pending.get(key)
    if slot:
        if is_metric_collapse(key, prev, slot.get("value")):
            if key == "c" and is_capacity_down_reset(prev, slot.get("value")):
                if is_capacity_down_reset(prev, raw) or is_collapsed_value(raw, "c"):
                    confirms = int(slot["confirms"]) + 1
                    if confirms >= CONFIRM_DEFAULT:
                        pending.pop(key, None)
                        return raw
                    pending[key] = {"value": raw, "confirms": confirms}
                    return prev
            if near_metric(prev, raw, key):
                pending.pop(key, None)
                return raw
            pending[key] = {"value": raw, "confirms": int(slot["confirms"]) + 1}
            return prev

        if near_metric(slot.get("value"), raw, key):
            confirms = int(slot["confirms"]) + 1
            if confirms >= CONFIRM_DEFAULT:
                pending.pop(key, None)
                return raw
            pending[key] = {"value": raw, "confirms": confirms}
            return prev
        if near_metric(prev, raw, key):
            pending.pop(key, None)
            return raw
        pending[key] = {"value": raw, "confirms": 1}
        return prev

    if metric_abrupt(prev, raw, key):
        # First capacity reading after unknown/null is a baseline, not a glitch.
        if key == "c" and prev is None and raw is not None:
            pending.pop(key, None)
            return raw
        pending[key] = {"value": raw, "confirms": 1}
        return prev

    return raw


def resolve_sample(
    pending: PendingState,
    prev: tuple[float | None, float | None, float | None] | None,
    incoming: tuple[Any, Any, Any],
) -> tuple[float | None, float | None, float | None]:
    raw = sanitize_raw(*incoming)
    if prev is None:
        pending.clear()
        return raw
    sample = (
        resolve_metric(pending, "v", prev[0], raw[0]),
        resolve_metric(pending, "i", prev[1], raw[1]),
        resolve_metric(pending, "c", prev[2], raw[2]),
    )
    if not pending:
        pending.clear()
    return sample


def _scrub_isolated(key: MetricKey, values: list[float | None]) -> None:
    """Original 1-sample sandwich: if neighbors agree and the middle jumped, hold left."""
    n = len(values)
    for i in range(1, n - 1):
        if near_metric(values[i - 1], values[i + 1], key) and metric_abrupt(values[i - 1], values[i], key):
            values[i] = values[i - 1]


def _scrub_metric(key: MetricKey, values: list[float | None]) -> None:
    _scrub_isolated(key, values)
    n = len(values)
    i = 1
    while i < n - 1:
        # Trickle plateaus (~90 mA) are not "stable" at the old 120 mA gate, but they
        # must still be eligible as the left side of a sandwich spike.
        if is_collapsed_value(values[i - 1], key):
            i += 1
            continue
        if not metric_abrupt(values[i - 1], values[i], key):
            i += 1
            continue

        run_start = i
        while i < n and metric_abrupt(values[run_start - 1], values[i], key):
            i += 1
        run_len = i - run_start
        if run_len < 1 or run_len > SCRUB_RUN_MAX or i >= n:
            continue

        left = values[run_start - 1]
        right = values[i]
        if is_collapsed_value(left, key) or not near_metric(left, right, key):
            continue

        collapse_run = all(is_collapsed_value(values[run_start + j], key) for j in range(run_len))
        fill = True
        if not collapse_run:
            for j in range(run_len):
                if not metric_abrupt(left, values[run_start + j], key):
                    fill = False
                    break
        if not fill:
            continue

        for j in range(run_len):
            t = (j + 1) / (run_len + 1)
            if left is not None and right is not None:
                values[run_start + j] = left + (right - left) * t
            else:
                values[run_start + j] = left


def hold_collapses(values: list[float | None], key: MetricKey) -> None:
    """Hold brief needles; keep sustained capacity down-resets (new process)."""
    n = len(values)
    i = 1
    while i < n:
        if not is_metric_collapse(key, values[i - 1], values[i]):
            i += 1
            continue
        left = values[i - 1]
        run_start = i
        while i < n and is_metric_collapse(key, left, values[i]):
            i += 1
        recovered = i < n and near_metric(left, values[i], key)
        # Capacity: only rewrite recovered sandwich needles. Trailing/sustained
        # down-resets are kept (resolve_metric already filters brief tips).
        hold = recovered if key == "c" else True
        if hold:
            for j in range(run_start, i):
                values[j] = left


def scrub_points(points: list[Point]) -> list[Point]:
    """Fill sandwich spikes and hold remaining collapses. Does not mutate `points`."""
    if len(points) < 2:
        return list(points)
    voltages = [p[2] for p in points]
    currents = [p[3] for p in points]
    capacities = [p[4] for p in points]
    if len(points) >= 3:
        _scrub_metric("v", voltages)
        _scrub_metric("i", currents)
        _scrub_metric("c", capacities)
    hold_collapses(voltages, "v")
    hold_collapses(currents, "i")
    hold_collapses(capacities, "c")
    return [
        (p[0], p[1], voltages[idx], currents[idx], capacities[idx]) for idx, p in enumerate(points)
    ]


class _ChannelSeries:
    __slots__ = ("t0_ms", "points", "pending")

    def __init__(self, t0_ms: int) -> None:
        self.t0_ms = t0_ms
        # (wall_ms, t_s, v, i, c)
        self.points: deque[Point] = deque(maxlen=MAX_POINTS)
        self.pending: PendingState = {}


class LiveSeriesStore:
    """Ring buffer of live samples. Not tied to hostname, origin, or browser."""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._channels: dict[int, _ChannelSeries] = {}
        # After Stop, skip ingest until the channel is seen idle (avoids 1–2 leftover ticks).
        self._hold_off: set[int] = set()
        # After Stop/Start, suppress stale high capacity until a fresh low baseline.
        self._capacity_fresh_since: dict[int, int] = {}

    def clear(self, channel: int, *, hold_off: bool = False) -> None:
        with self._lock:
            ch = int(channel)
            self._channels.pop(ch, None)
            self._capacity_fresh_since[ch] = _now_ms(None)
            if hold_off:
                self._hold_off.add(ch)
            else:
                self._hold_off.discard(ch)

    def clear_all(self) -> None:
        with self._lock:
            self._channels.clear()
            self._hold_off.clear()
            self._capacity_fresh_since.clear()

    def _arm_capacity_locked(
        self, ch: int, cap: float | None, *, running: bool, now_ms: int
    ) -> float | None:
        since = self._capacity_fresh_since.get(ch)
        if since is None:
            return cap
        if not running:
            return None
        if (
            cap is None
            or not is_stable_value(cap, "c")
            or now_ms - since >= CAPACITY_FRESH_TIMEOUT_MS
        ):
            self._capacity_fresh_since.pop(ch, None)
            return cap
        return None

    def ingest(
        self,
        channels: list[dict[str, Any]] | None,
        measurements: list[dict[str, Any]] | None,
        now_ms: int | None = None,
    ) -> None:
        ts = _now_ms(now_ms)
        with self._lock:
            self._prune_locked(ts)
            if not channels:
                return
            ch_list = [{**c, "channel": idx} for idx, c in enumerate(channels)]
            meas_in = list(measurements or [])
            if len(meas_in) == len(ch_list):
                meas_list = [{**m, "channel": idx} for idx, m in enumerate(meas_in)]
            else:
                meas_list = meas_in
            meas_by_ch = {int(m.get("channel", i)): m for i, m in enumerate(meas_list)}

            for c in ch_list:
                ch = int(c["channel"])
                if ch in self._hold_off:
                    if is_idle_channel(c):
                        self._hold_off.discard(ch)
                    continue
                if is_idle_channel(c):
                    continue
                m = meas_by_ch.get(ch, {})
                series = self._channels.get(ch)
                if series is None:
                    series = _ChannelSeries(ts)
                    self._channels[ch] = series
                t = (ts - series.t0_ms) / 1000.0
                prev = None
                if series.points:
                    last = series.points[-1]
                    prev = (last[2], last[3], last[4])
                v_raw, i_raw, c_raw = sanitize_raw(
                    m.get("voltage_V"), m.get("current_mA"), m.get("capacity_mAh")
                )
                c_raw = self._arm_capacity_locked(ch, c_raw, running=True, now_ms=ts)
                v, i, cap = resolve_sample(
                    series.pending,
                    prev,
                    (v_raw, i_raw, c_raw),
                )
                series.points.append((ts, t, v, i, cap))
                self._scrub_tail_locked(series)

    def snapshot(self, now_ms: int | None = None) -> dict[str, Any]:
        ts = _now_ms(now_ms)
        with self._lock:
            self._prune_locked(ts)
            out: dict[str, Any] = {}
            for ch, series in self._channels.items():
                cleaned = scrub_points(list(series.points))
                series.points.clear()
                series.points.extend(cleaned)
                out[str(ch)] = {
                    "t0": series.t0_ms,
                    "points": [{"t": t, "v": v, "i": i, "c": cap} for _wall, t, v, i, cap in cleaned],
                }
            return {"channels": out}

    def _scrub_tail_locked(self, series: _ChannelSeries) -> None:
        n = len(series.points)
        if n < 3:
            return
        window = min(n, SCRUB_RUN_MAX * 2 + 3)
        pts = list(series.points)
        head, tail = pts[:-window], pts[-window:]
        series.points.clear()
        series.points.extend(head)
        series.points.extend(scrub_points(tail))

    def _prune_locked(self, now_ms: int) -> None:
        cutoff = now_ms - _MAX_AGE_MS
        drop: list[int] = []
        for ch, series in self._channels.items():
            pts = series.points
            while pts and pts[0][0] < cutoff:
                pts.popleft()
            if not pts:
                drop.append(ch)
        for ch in drop:
            self._channels.pop(ch, None)
